### ISSUE TITLE
Add a PID file so only one instance of aw-watcher-terminal is needed

### ISSUE THREAD

**Issue by johan-bjareholt** [Otto-AA/aw-watcher-terminal#1]
Seems to me like you currently start a aw-watcher-terminal instance for each terminal which is opened. Since you are using one-way communication via fifo-files multiple instances should not be needed.

I am suggesting to use the activity watch data dir (.local/share/activitywatch/aw-watcher-terminal on linux, accessible via aw-core.dirs.get_data_dir) for both the fifo-file and a pid-file. Then we can check every time aw-watcher-terminal starts if a aw-watcher-terminal process is already running and if it is we can simply just utilize that instance and its fifo-file instead. This should also fix issues on multi-user systems when you have the fifo file in /tmp.

**Comment by Otto-AA** [comment]
Actually, aw-watcher-terminal.py is only started once by the user (as described in the [installation process](https://github.com/Otto-AA/aw-watcher-terminal#installation)). Maybe you mixed it up with aw-watcher-bash(.sh), which is started for every command executed?

> I am suggesting to use the activity watch data dir (.local/share/activitywatch/aw-watcher-terminal on linux, accessible via aw-core.dirs.get_data_dir) for both the fifo-file and a pid-file.

Seems like a good idea to put the fifo-file in there :+1: 
Also the pid-file could be useful for aw-watcher-bash.sh. Currently it just attempts to write to the pipe and gives up after five seconds. With that we could check if aw-watcher-terminal is actually running and only then attempt writing to the pipe.

**Comment by Otto-AA** [comment]
> I am suggesting to use the activity watch data dir (.local/share/activitywatch/aw-watcher-terminal on linux, accessible via aw-core.dirs.get_data_dir) for both the fifo-file and a pid-file.

Do you have an idea, on how to get this path in the bash file? One not so nice way, would be to tell the user to find out the path (by looking at the loggings from aw-watcher-terminal or smth. similar) and pass it as param in the .profile file.

**Comment by johan-bjareholt** [comment]
> Actually, aw-watcher-terminal.py is only started once by the user (as described in the installation process). Maybe you mixed it up with aw-watcher-bash(.sh), which is started for every command executed?

Aha, read that wrong I guess.

>  # Start aw-watcher-terminal.py
>  python3 ./.local/bin/aw-watcher-terminal.py

This part is still quite troublesome though, we don't really want to have to manually start aw-watcher-terminal and have it as a child process of the terminal in case that specific terminal is shut down.
My idea with the PID file was to have it running as a daemon and check at every startup if it is already running, and if not start aw-watcher-terminal as a daemon.

> Also the pid-file could be useful for aw-watcher-bash.sh. Currently it just attempts to write to the pipe and gives up after five seconds. With that we could check if aw-watcher-terminal is actually running and only then attempt writing to the pipe.

Well, it should be fine to write to the pipe file even if aw-watcher-terminal is not running since in that case the events will just queue up and be sent the next time aw-watcher-terminal is started. I don't think it's possible to stall when writing to a fifo file so there is no point in having a timeout, but I could be wrong.

> Do you have an idea, on how to get this path in the bash file? One not so nice way, would be to tell the user to find out the path (by looking at the loggings from aw-watcher-terminal or smth. similar) and pass it as param in the .profile file.

I think you can just hardcode ~/.local/share/activitywatch/aw-watcher-terminal, should be fine in almost all cases. Not sure if it will be correct on macOS though.

**Comment by johan-bjareholt** [comment]
> I think you can just hardcode ~/.local/share/activitywatch/aw-watcher-terminal, should be fine in almost all cases. Not sure if it will be correct on macOS though.

Correction, a 100% reliable solution would be `${XDG_DATA_HOME:-$HOME/.local/share}/activitywatch/aw-watcher-terminal`

**Comment by Otto-AA** [comment]
> Well, it should be fine to write to the pipe file even if aw-watcher-terminal is not running since in that case the events will just queue up and be sent the next time aw-watcher-terminal is started. I don't think it's possible to stall when writing to a fifo file so there is no point in having a timeout, but I could be wrong.

When attempting to write to a named pipe (e.g. `echo '--command "test"' > "$pipe_path"`), the execution of this command isn't finished until aw-watcher-terminal reads it ("Normally, opening the FIFO blocks until the other end is opened also." - [fifo man page](https://linux.die.net/man/7/fifo)). Therefore the process would be kept open until the aw-watcher-terminal reads the pipe, which can lead to hundreds of open stalling bash processes.

To solve this, I've pushed it to a background process, and made a timeout (5s) which kills the background process if still existing.

If you have a better solution, I'd be glad to hear it :)
The fifo man page also states, that "A process can open a FIFO in nonblocking mode", yet I haven't found a way to do that in bash.

**Comment by johan-bjareholt** [comment]
> When attempting to write to a named pipe (e.g. echo '--command "test"' > "$pipe_path"), the execution of this command isn't finished until aw-watcher-terminal reads it ("Normally, opening the FIFO blocks until the other end is opened also." - fifo man page). Therefore the process would be kept open until the aw-watcher-terminal reads the pipe, which can lead to hundreds of open stalling bash processes.
> 
> To solve this, I've pushed it to a background process, and made a timeout (5s) which kills the background process if still existing.

Aha, that's good enough then.

Even if we would get that working with bash the fifo could still fill up and fail to write after a while: (man 7 pipe)

> O_NONBLOCK enabled, n <= PIPE_BUF
>        If there is room to write n bytes to the pipe, then write(2) succeeds immediately, writing all n bytes; otherwise write(2) fails, with errno set to EAGAIN.

**Comment by Otto-AA** [comment]
> My idea with the PID file was to have it running as a daemon and check at every startup if it is already running, and if not start aw-watcher-terminal as a daemon.

I think, the best outcome would be, that it also can be autostarted with aw-qt. (I guess, [this](https://github.com/ActivityWatch/aw-qt/blob/master/aw_qt/manager.py#L112) todo needs to be done beforehand)
As far as I have seen, this would require the watcher to be compiled into an executable which could be done with PyInstaller.

Checking on every terminal startup seems like a good idea to me. I'd prefer to keep this logic separate from aw-watcher-bash (or other shell specific watchers) and rather implement it somehow in aw-watcher-terminal and/or aw-qt, but if that's not possible we can do it this way.

But to be honest, I don't really have a clue how to do any of the above, so help (e.g. links to related documentation or examples) would be appreciated :)

**Comment by johan-bjareholt** [comment]
> I think, the best outcome would be, that it also can be autostarted with aw-qt. (I guess, this todo needs to be done beforehand)
> As far as I have seen, this would require the watcher to be compiled into an executable which could be done with PyInstaller.

You are correct, that is probably a better idea.

On Linux it doesn't have to be an executable, just need to:
- Have the python dependencies installed
- the python program named aw-watcher-terminal
- shebang "#!/usr/bin/env python3" at top of file
- executable permissions on the file

Then you can for now just hardcode in "aw-watcher-terminal" to that _possibe_modules list for now. Finally just have aw-watcher-terminal in your $PATH and it should hopefully work.

**Comment by Otto-AA** [comment]
Is there still something left open in this issue?
If not, I'd close this issue and if you have other suggestions/questions/... you can open another issue :)

**Comment by johan-bjareholt** [comment]
Don't think so :)
